kolena/workflow/_datatypes.py

```diff
@@ -44,10 +45,82 @@
 T = TypeVar("T", bound="DataObject")
 
 
+def _double_under(input: str) -> bool:
+    return input.startswith("__") and input.endswith("__")
+
+
+def _allow_extra(cls: Type[T]) -> bool:
+    # `pydantic.dataclasses.is_built_in_dataclass` would have false-positive when a stdlib-dataclass decorated
+    # class extends a pydantic dataclass
+    return "__pydantic_model__" in vars(cls) and cls.__pydantic_model__.Config.extra == Extra.allow
+
+
+# used to track data_type string -> TypedDataObject
+_DATA_TYPE_MAP = {}
+
+
+def _get_full_type(obj: "TypedDataObject") -> str:
+    data_type = obj._data_type()
+    return f"{data_type._data_category()}/{data_type.value}"
+
+
+def _get_data_type(name: str) -> Optional[Type["TypedDataObject"]]:
+    return _DATA_TYPE_MAP.get(name, None)
+
+
+# used for TypedBaseDataObject to register themselves to be used in dataclass extra fields deserialization
+def _register_data_type(cls) -> None:
+    full_name = _get_full_type(cls)
+    # leverage class inheritance order, only keep base classes of a datatype
+    if full_name not in _DATA_TYPE_MAP:
+        _DATA_TYPE_MAP[full_name] = cls
+
+
+def _deserialize_typed_dataobject(value: Dict[Any, Any]) -> Any:
+    data_type = _get_data_type(value[DATA_TYPE_FIELD])
+    if data_type is None:
+        return value
+
+    # 'data_type' is not a real member
+    value.pop(DATA_TYPE_FIELD)
+    return data_type._from_dict(value)
+
+
+# best effort to deserialize typed data objects in dataclass extra fields
+# note: since a "data_type" string could map to multiple classes through inheritance, only base case would be used.
+def _try_deserialize_typed_dataobject(value: Any) -> Any:
+    if isinstance(value, list):
+        # only attempt deserialization when it is likely this is a list of typed data objects
+        if value and isinstance(value[0], dict) and DATA_TYPE_FIELD in value[0]:
+            return [_try_deserialize_typed_dataobject(val) for val in value]
+    elif isinstance(value, dict):
+        if DATA_TYPE_FIELD in value:
+            return _deserialize_typed_dataobject(value)
+
+    return value
+
+
 @dataclass(frozen=True, config=ValidatorConfig)
 class DataObject(metaclass=ABCMeta):
     """The base for various objects in `kolena.workflow`."""
 
+    def __str__(self):
```

Comment:
Implemented custom `__str__` to include extra fields. My original oversight was that I focused on `__repr__` instead. `dataclass` doesn't generate `__str__`, only `__repr__`, as such provides us an window for `print` customization.

cc @gordonhart @munkyshi @yifanwu-kolena for another review. See unit tests for details, specifically `test__data_object__extras_allow`